### PR TITLE
Adjusting the QP encoding to deal with quantification over finite types

### PR DESCRIPTION
**Pull request by marcoeilers**
Adapting the QP encoding to soundly deal with quantification over finite types, as described here: https://pm.inf.ethz.ch/publications/Schwerhoff16.pdf#page=122

Fixes issue #342

### CODE REVIEW

File: src/main/scala/rules/QuantifiedChunkSupport.scala [hunk 1]
```diff
@@ -103,18 +108,21 @@ trait QuantifiedChunkSupport extends SymbolicExecutionRules {
     *     the partial inverses are defined
     *
     * The following definitional axioms will be returned, in addition to the fresh
-    * inverse functions `inv_1`, ..., `inv_n`:
+    * inverse functions `inv_1`, ..., `inv_n` and the respective image functions:
     *
     *   forall x_1: T_1, ..., x_n: T_n :: {fct(x_1, ..., x_n)}
     *     c(x_1, ..., x_n) ==>
-    *          inv_1(fct(x_1, ..., x_n)) == x_1
+    *          inv_1(fct(x_1, ..., x_n)) == x_1 && img_1(fct(x_1, ..., x_n))
```
Comment (mschwerhoff) [inline src/main/scala/rules/QuantifiedChunkSupport.scala:115]: Are `img_i` functions of boolean types?

File: src/main/scala/rules/QuantifiedChunkSupport.scala [hunk 2]
```diff
@@ -103,18 +108,21 @@ trait QuantifiedChunkSupport extends SymbolicExecutionRules {
...
     *       && ...
-    *       && inv_n(fct(x_1, ..., x_n)) == x_n
+    *       && inv_n(fct(x_1, ..., x_n)) == x_n && img_n(fct(x_1, ..., x_n))
     *
     *   forall r: R :: {inv_1(r), ..., inv_n(r)}
-    *     c(inv_1(r), ..., inv_n(r)) ==>
+    *     c(inv_1(r), ..., inv_n(r)) && img_1(r) && ... && img_n(r) ==>
     *       fct(inv_1(r), ..., inv_n(r)) == r
     *
+    *  For all i where x_i is of type Bool or Int, we do not generate the img_i constraints in
```
Comment (mschwerhoff) [inline src/main/scala/rules/QuantifiedChunkSupport.scala:123]: Why do `Bool` and `Int` have the same domain as `Ref`? I'm confused.

File: src/main/scala/rules/QuantifiedChunkSupport.scala [hunk 3]
```diff
@@ -1589,6 +1603,19 @@ object quantifiedChunkSupporter extends QuantifiedChunkSupport {
...
+      if (qvar.sort != sorts.Int && qvar.sort != sorts.Ref) {
+        // Types known to be infinite, thus there is no need to constrain the domain using image functions.
+        val imgFun = v.decider.fresh("img", (additionalInvArgs map (_.sort)) ++ invertibles.map(_.sort), sorts.Bool)
+        val img = (ts: Seq[Term]) => App(imgFun, additionalInvArgs ++ ts)
+
+        imageFunctions(idx) = imgFun
+        imagesOfFcts(idx) = img(invertibles)
+        imagesOfCodomains(idx) = img(codomainQVars)
+      } else {
+        imagesOfFcts(idx) = True()
```
Comment (mschwerhoff) [inline src/main/scala/rules/QuantifiedChunkSupport.scala:1616]: A comment `// imageFunctions(idx) remains null` couldn't hurt, and would help understanding the [filter in line 1693](https://github.com/viperproject/silicon/pull/666/files#diff-aed5f0e36948a8a0ce2bcd027936b8cbe7fbbb8fc68bf9ef2f2ea10ba8c82f2dR1693).

**Review by mschwerhoff — APPROVED** [review]
Approved - assuming my confusion isn't caused by any issues.

File: src/main/scala/rules/QuantifiedChunkSupport.scala (hunk 1 above)
Comment (marcoeilers) [reply src/main/scala/rules/QuantifiedChunkSupport.scala:115]: Yes.

File: src/main/scala/rules/QuantifiedChunkSupport.scala (hunk 2 above)
Comment (marcoeilers) [reply src/main/scala/rules/QuantifiedChunkSupport.scala:123]: Oof, that's a typo. I mean `Int` and `Ref`.

**Pull request merged**